tools/libopaeuio/opaeuio.c

```diff
@@ -0,0 +1,356 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <glob.h>
+#include <dirent.h>
+
+#include <opae/uio.h>
+
+#define __SHORT_FILE__                                    \
+({                                                        \
+	const char *file = __FILE__;                      \
+	const char *p = file;                             \
+	while (*p)                                        \
+		++p;                                      \
+	while ((p > file) && ('/' != *p) && ('\\' != *p)) \
+		--p;                                      \
+	if (p > file)                                     \
+		++p;                                      \
+	p;                                                \
+})
+
+#define ERR(format, ...)                                \
+fprintf(stderr, "%s:%u:%s() **ERROR** [%s] : " format , \
+	__SHORT_FILE__, __LINE__, __func__, strerror(errno), ##__VA_ARGS__)
+
+STATIC int opae_uio_read_sysfs_uint64(const char *path, uint64_t *puint)
+{
+	FILE *fp;
+
+	fp = fopen(path, "r");
+	if (!fp) {
+		ERR("failed to fopen(\"%s\", \"r\")\n", path);
+		return 1;
+	}
+
+	if (fscanf(fp, "%lx", puint) != 1) {
+		ERR("fscanf() failed\n");
+		fclose(fp);
+		return 2;
+	}
+
+	fclose(fp);
+
+	return 0;
+}
+
+STATIC struct opae_uio_device_region *
+opae_uio_device_region_create(uint32_t index,
+			      uint8_t *ptr,
+			      size_t page_offset,
+			      size_t size)
+{
+	struct opae_uio_device_region *r;
+	r = malloc(sizeof(*r));
+	if (r) {
+		r->region_index = index;
+		r->region_ptr = ptr;
+		r->region_page_offset = page_offset;
+		r->region_size = size;
+		r->next = NULL;
+	}
+	return r;
+}
+
+STATIC void
+opae_uio_device_region_destroy(struct opae_uio_device_region *r)
+{
+	while (r) {
+		struct opae_uio_device_region *trash = r;
+		r = r->next;
+		if (trash->region_ptr != MAP_FAILED) {
+			if (munmap(trash->region_ptr, trash->region_size) < 0)
+				ERR("munmap() failed\n");
+		}
+		free(trash);
+	}
+}
+
+STATIC void opae_uio_destroy(struct opae_uio *u)
+{
+	opae_uio_device_region_destroy(u->regions);
+	u->regions = NULL;
+
+	if (u->device_fd >= 0) {
+		close(u->device_fd);
+		u->device_fd = -1;
+	}
+}
+
+STATIC struct opae_uio_device_region *
+opae_uio_enum_region(int fd, char *sysfs_path)
+{
+	uint32_t index = 0;
+	uint8_t *ptr = NULL;
+	uint64_t page_offset = 0;
+	uint64_t size = 0;
+	char *p;
+	char *endptr;
+	char path[OPAE_UIO_PATH_MAX];
+	uint32_t page_size = 0;
+
+	// The region index is encoded in the file name component.
+	p = strrchr(sysfs_path, '/');
+	if (!p) {
+		ERR("invalid map path: \"%s\"\n", sysfs_path);
+		return NULL;
+	}
+
+	endptr = NULL;
+	index = strtoul(p + 4, &endptr, 10);
```

Comment:
p + 4 means ? why +4

Reply:
Because the digits start at offset 4 from the last '/' character.
```c
01234
/map0
```